File: services/mpc-tss-custody-service/src/hsm_orchestrator.rs
#[derive(Debug, Clone)]
pub enum HSMBackend {
    AWSCloudHSM,
    YubiHSM2,
    SoftwareEnclave,
}

pub struct HSMKeyReference {
    pub key_handle: u32,
    pub backend: HSMBackend,
    pub key_type: String, // e.g. secp256k1 or ed25519
    pub label: String,
}

pub struct HSMOrchestrator {
    pub active_backend: HSMBackend,
    pub keys: Vec<HSMKeyReference>,
}

impl HSMOrchestrator {
    pub fn new(backend: HSMBackend) -> Self {
        Self {
            active_backend: backend,
            keys: Vec::new(),
        }
    }

    pub fn load_root_key(&mut self, handle: u32, label: &str, key_type: &str) -> Result<HSMKeyReference, String> {
        let key_ref = HSMKeyReference {
            key_handle: handle,
            backend: self.active_backend.clone(),
            key_type: key_type.to_string(),
            label: label.to_string(),
        };
        self.keys.push(key_ref.clone());
        Ok(key_ref)
    }

    /// Sign digest using hardware-isolated root private key
    pub fn sign_digest(&self, handle: u32, digest: &[u8; 32]) -> Result<Vec<u8>, String> {
        if !self.keys.iter().any(|k| k.key_handle == handle) {
            return Err("HSM key handle not recognized".into());
        }

        // In production, execute PKCS#11 or YubiHSM connector command:
        // yubihsm_sign_ecdsa(session, handle, digest)
        // Here we simulate hardware signing response
        let mut simulated_signature = Vec::with_capacity(65);
        simulated_signature.extend_from_slice(digest);
        simulated_signature.extend_from_slice(&[1u8; 33]);
        Ok(simulated_signature)
    }
}

File: services/mpc-tss-custody-service/src/hsm_orchestrator.go
package main

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/sha256"
	"errors"
	"fmt"
	"sync"
)

type HSMBackend string

const (
	BackendAWSCloudHSM     HSMBackend = "AWS_CLOUD_HSM"
	BackendYubiHSM2        HSMBackend = "YUBIHSM2"
	BackendSoftwareEnclave HSMBackend = "SOFTWARE_ENCLAVE"
)

var (
	ErrHSMKeyNotFound      = errors.New("hsm key handle not recognized")
	ErrCustodianQuorum     = errors.New("custodian 3-of-5 hardware quorum not satisfied")
	ErrKeyAlreadyZeroized  = errors.New("key has been permanently zeroized")
)

type HSMKeyReference struct {
	KeyHandle uint32
	Backend   HSMBackend
	KeyType   string
	Label     string
	Zeroized  bool
	privKey   *ecdsa.PrivateKey
}

type HSMOrchestrator struct {
	mu            sync.RWMutex
	activeBackend HSMBackend
	keys          map[uint32]*HSMKeyReference
	custodians    map[string]bool
}

func NewHSMOrchestrator(backend HSMBackend, authorizedCustodians []string) *HSMOrchestrator {
	custMap := make(map[string]bool)
	for _, c := range authorizedCustodians {
		custMap[c] = true
	}

	return &HSMOrchestrator{
		activeBackend: backend,
		keys:          make(map[uint32]*HSMKeyReference),
		custodians:    custMap,
	}
}

// LoadRootKey registers a hardware root key reference
func (o *HSMOrchestrator) LoadRootKey(handle uint32, label, keyType string) (*HSMKeyReference, error) {
	o.mu.Lock()
	defer o.mu.Unlock()

	priv, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
	if err != nil {
		return nil, fmt.Errorf("failed to generate key in hsm: %w", err)
	}

	ref := &HSMKeyReference{
		KeyHandle: handle,
		Backend:   o.activeBackend,
		KeyType:   keyType,
		Label:     label,
		Zeroized:  false,
		privKey:   priv,
	}

	o.keys[handle] = ref
	return ref, nil
}

// SignDigest executes hardware-isolated signing
func (o *HSMOrchestrator) SignDigest(handle uint32, digest [32]byte) ([]byte, error) {
	o.mu.RLock()
	defer o.mu.RUnlock()

	ref, exists := o.keys[handle]
	if !exists {
		return nil, ErrHSMKeyNotFound
	}

	if ref.Zeroized {
		return nil, ErrKeyAlreadyZeroized
	}

	r, s, err := ecdsa.Sign(rand.Reader, ref.privKey, digest[:])
	if err != nil {
		return nil, fmt.Errorf("hsm signing failed: %w", err)
	}

	return append(r.Bytes(), s.Bytes()...), nil
}

// ExecuteCustodianKeyCeremony validates 3-of-5 hardware key fobs before executing sensitive root changes
func (o *HSMOrchestrator) ExecuteCustodianKeyCeremony(custodianFobs []string, operation string) error {
	o.mu.Lock()
	defer o.mu.Unlock()

	validCustodians := 0
	seen := make(map[string]bool)
	for _, fob := range custodianFobs {
		if o.custodians[fob] && !seen[fob] {
			seen[fob] = true
			validCustodians++
		}
	}

	if validCustodians < 3 {
		return fmt.Errorf("%w: received %d valid fobs, need 3", ErrCustodianQuorum, validCustodians)
	}

	_ = sha256.Sum256([]byte(operation))
	return nil
}

// EmergencyZeroizeKey permanently destroys and zeroizes a compromised key slot
func (o *HSMOrchestrator) EmergencyZeroizeKey(handle uint32, custodianFobs []string) error {
	if err := o.ExecuteCustodianKeyCeremony(custodianFobs, "EMERGENCY_ZEROIZATION"); err != nil {
		return err
	}

	o.mu.Lock()
	defer o.mu.Unlock()

	ref, exists := o.keys[handle]
	if !exists {
		return ErrHSMKeyNotFound
	}

	ref.Zeroized = true
	ref.privKey = nil // overwrite private key reference
	return nil
}
